drivers/gpio/gpio_stm32.c

```diff
@@ -561,6 +581,63 @@ static const struct gpio_driver_api gpio_stm32_driver = {
 	.manage_callback = gpio_stm32_manage_callback,
 };
 
+#ifdef CONFIG_PM_DEVICE
+static uint32_t gpio_stm32_get_power_state(const struct device *dev)
+{
+	struct gpio_stm32_data *data = dev->data;
+
+	return data->power_state;
+}
+
+int gpio_stm32_set_power_state(const struct device *dev,
+					      uint32_t new_state)
+{
+	struct gpio_stm32_data *data = dev->data;
+	int ret = 0;
+
+	if (new_state == DEVICE_PM_ACTIVE_STATE) {
+		ret = gpio_stm32_clock_request(dev, true);
+	} else if (new_state == DEVICE_PM_LOW_POWER_STATE) {
+		ret = gpio_stm32_clock_request(dev, false);
+	}
+
+	data->power_state = new_state;
+
+	return ret;
+}
+
+static int gpio_stm32_device_pm_ctrl(const struct device *dev,
+				     uint32_t ctrl_command,
+				     void *context, device_pm_cb cb, void *arg)
+{
+	struct gpio_stm32_data *data = dev->data;
+	uint32_t new_state;
+	int ret = 0;
+
+	switch (ctrl_command) {
+	case DEVICE_PM_SET_POWER_STATE:
+		new_state = *((const uint32_t *)context);
+		if (new_state != data->power_state) {
+			ret = gpio_stm32_set_power_state(dev, new_state);
+		}
+		break;
+	case DEVICE_PM_GET_POWER_STATE:
+		*((uint32_t *)context) = gpio_stm32_get_power_state(dev);
+		break;
+	default:
+		ret = -EINVAL;
+
+	}
+
+	if (cb) {
```

Comment:
`gpio_stm32_set_power_state` is always sync ?

Reply:
`gpio_stm32_set_power_state` is always sync, yes. But tbh, I'm not clear on that part. What is the expectation here?

Reply:
> @ceolin Thanks for the review (I'll fix what could be fixed). But I would also have liked to have feedback on the points I reported in the comment [above](https://github.com/zephyrproject-rtos/zephyr/pull/32788#issuecomment-801166195)

Regarding the first point, I think the combination of system constraints and `device_pm_put/get` suppress the need of that API.
The second point is not clear for me what is the problem, if it is just a bug or some architectural problem. Need to debug to understand better.
The third point I'm not sure what we should do, maybe devices should do what they do in the system initialization ? or maybe just restore from what they were before sleep ?